d3b-center/image-deid-etl
# Store processed UUIDs with SQLAlchemy

## Overview

I made the following changes to get us closer to storing all of the ETL's state in a transactional datastore:

- Create an `etl/database` module for interfacing with the database via SQLAlchemy.
- Everywhere that used to `json.dumps`/`json.load` now queries the database for the same information.
- Add three new subcommands:
  - `initdb` initializes the empty database.
  - `importuuids` imports processed UUIDs from a JSON file.
  - `check` replaces `--check_orthanc`.
- Move all of the ETL-related steps into a `run_pipeline` subcommand. 
  - This is a small step towards #11.

Connects #5 

### Checklist


### Notes

When building any application, it's important to maintain as much consistency as possible between development and production. For example, my typical pattern when developing Python applications is to containerize the application and specify it, along with any supporting services like PostgreSQL, in a  `docker-compose.yml` file. This pattern is a great way to achieve consistency with minimal divergence between two environments. Both environments will use the same container and database. And, the application and database can talk to each other over the network namespace provided by Docker.

I decided to take an experimental approach with this project by using Nix to manage the development environment. With Nix, it's much easier to manage dependencies, which seemed like a win when it became clear we'd have to vendor the Flywheel CLI and compile dcm2niix from source. However, it's not as seamless when you want to integrate with something like Docker Compose. 

`docker-compose up` won't automatically build an image with Nix. Instead, we'd have to add a step to `scripts/update` that bakes the base image before folks can interact with the project.

I think we should do this down the road, but I didn't want to add new dependencies at this stage, so I decided to use SQLite in the development environment, which doesn't require any supporting services. The downside is we're diverging from what will run in production. However, since we're using SQLAlchemy to interface with the database, and our schema is relatively simple, our application should be pluggable between the two.

**Note:** https://12factor.net/dev-prod-parity is a good read on the merits of minimizing consistency.

## Testing Instructions

First, run `update` to ensure the local SQLite database is initialized:

```console
$ ./scripts/update
. . .
Initializing database schema...
Success!
```

Next, remove the following Orthanc UUIDs from the tail end of `orthanc_study_uuids.json`:

- `f90cd746-10017fbb-44472a28-391f1a52-3186cf1a`
- `d1b90c78-d9b579f2-a4664679-7ede472b-978d73a2`

Now, use the new `importuuids` command to import these UUIDs into the new database:

```console
$ ./image_deid_etl importuuids orthanc_study_uuids.json 
Imported 3055 new UUIDs.
```

Finally, we can run `test` to process the two studies we just removed:

```console
$ ./scripts/test 
Checking for new UUIDs.
Found new UUID(s) Orthanc, beginning to download 2 new studies (max_studies = 40).
. . .
DONE PROCESSING STUDIES
There are files to check in: cbtn/chop/NIfTIs_short_json/
Updating list of UUIDs...
```

And now, confirm that our local database is in-sync with Orthanc:

```console
$ ./image_deid_etl check
No new UUIDs found on Orthanc
```
